crates/red_knot/src/symbols.rs

```diff
@@ -109,11 +137,31 @@ impl SymbolTable {
         &self.scopes_by_id[self.root_scope_id()]
     }
 
-    pub(crate) fn add_symbol_to_scope(&mut self, scope_id: ScopeId, name: &str) -> SymbolId {
-        let mut hasher = FxHasher::default();
-        name.hash(&mut hasher);
-        let hash = hasher.finish();
+    pub(crate) fn symbols_for_scope(&self, scope_id: ScopeId) -> SymbolIterator {
+        let scope = &self.scopes_by_id[scope_id];
+        SymbolIterator {
+            table: self,
+            ids: scope.symbols_by_name.keys().cloned().collect(),
```

Comment:
Ideally, Iterators don't allocate. 

We can rewrite the iterator to be generic over any `Iterator` that returns a `SymbolId`.


```rust
struct SymbolIterator<'a, I> {
    table: &'a SymbolTable,
    ids: I,
}

impl<'a, I> Iterator for SymbolIterator<'a, I>
where
    I: Iterator<Item = SymbolId>,
{
    type Item = &'a Symbol;

    fn next(&mut self) -> Option<Self::Item> {
        let id = self.ids.next()?;
        Some(&self.table.symbols_by_id[id])
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        self.ids.size_hint()
    }
}

impl<'a, I> FusedIterator for SymbolIterator<'a, I> where
    I: Iterator<Item = SymbolId> + FusedIterator
{
}

impl<'a, I> ExactSizeIterator for SymbolIterator<'a, I> where
    I: Iterator<Item = SymbolId> + ExactSizeIterator
{
}

impl<'a, I> DoubleEndedIterator for SymbolIterator<'a, I>
where
    I: DoubleEndedIterator + Iterator<Item = SymbolId>,
{
    fn next_back(&mut self) -> Option<Self::Item> {
        let id = self.ids.next_back()?;
        Some(&self.table.symbols_by_id[id])
    }
}
```

You can then use it with

```rust
    pub(crate) fn symbols_for_scope(
        &self,
        scope_id: ScopeId,
    ) -> SymbolIterator<std::iter::Copied<hashbrown::hash_map::Keys<SymbolId, ()>>> {
        let scope = &self.scopes_by_id[scope_id];
        SymbolIterator {
            table: self,
            ids: scope.symbols_by_name.keys().copied(),
        }
    }
```

It could make sense to define an alias for it. Note. the old implementation poped the IDs from the back (not sure why). The new implementation consumes the Ids in the order returned by `keys`. 

General advice when implementing iterators:

* Override `size_hint` if you can estimate the number of elements. It is used when e.g. collecting into a vec
* Implement `FusedIterator` when it is guaranteed that the iterator returns no `Some` elements after it returned one `None` element. 
* Implement `DoubleEndedIterator` and `ExactSizeIterator` if possible.